core-blocks/link-preview/index.js

```diff
@@ -0,0 +1,252 @@
+			this.setState( { mode: this.MODES.FETCH } );
+			apiRequest( { path: apiURL } )
+				.then(
+					( obj ) => {
+						if ( this.unmounting ) {
+							return;
+						}
+						this.props.setAttributes( obj );
+						this.setState( { mode: this.MODES.EDITING } );
+					},
+					() => {
+						this.setState( { mode: this.MODES.CANT_FETCH } );
+					}
+				);
+			return;
+		}
+
+		this.setState( { mode: this.MODES.EDITING } );
+	}
+
+	componentWillUnmount() {
+		// can't abort the promise, so let it know we will unmount
+		this.unmounting = true;
+	}
+
+	setURL( event ) {
+		const { url } = this.state;
+		this.props.setAttributes( { url } );
+		if ( event ) {
+			event.preventDefault();
+		}
+	}
+
+	render() {
+		const { attributes, isSelected } = this.props;
+		const { mode, url, selectedImage, images } = this.state;
+		const { FETCH, CANT_FETCH, INPUT_URL } = this.MODES;
+		const isEditing = mode === INPUT_URL || mode === CANT_FETCH || mode === FETCH;
+		const hasMultipleImages = images.length > 1;
+		const label = __( 'Link preview' );
+
+		if ( isEditing ) {
+			return (
+				<Fragment>
+					<Placeholder label={ label } className="wp-block-embed">
+						<form onSubmit={ this.setURL }>
+							{ mode !== FETCH && (
+								<Fragment>
+									<input
+										type="url"
+										value={ url || '' }
+										className="components-placeholder__input"
+										aria-label={ label }
+										placeholder={ __( 'Enter URL here…' ) }
+										onChange={ ( event ) => this.setState( { url: event.target.value } ) } />
+									<Button
+										isLarge
+										type="submit">
+										{ __( 'Preview' ) }
+									</Button>
+								</Fragment>
+							) }
+							{ mode === FETCH && (
+								<div className="wp-block-embed-link-preview is-loading">
+									<Spinner />
+									<p>{ __( 'Generating preview…' ) }</p>
+								</div>
+							) }
+							{ mode === CANT_FETCH && <p className="components-placeholder__error">{ __( 'Sorry, we could not generate a preview for that URL.' ) }</p> }
+						</form>
+					</Placeholder>
+				</Fragment>
+			);
+		}
+
+		return (
+			<div className="wp-block-embed-link-preview">
+				<div className="wp-block-embed-link-preview__textinfo">
+					<p><a href={ attributes.url }><RawHTML>{ attributes.title }</RawHTML></a></p>
+					<p className="wp-block-embed-link-preview__description"><RawHTML>{ attributes.description }</RawHTML></p>
+				</div>
+				{ selectedImage && (
+					<div className="wp-block-embed-link-preview__image">
+						<div className="wp-block-embed-link-preview__image__selected">
+							<img src={ selectedImage.src } alt="" />
+						</div>
+						{ isSelected &&
+						<div className="wp-block-embed-link-preview__image__tools">
+							{ hasMultipleImages && <IconButton aria-label="Previous image" icon="arrow-left-alt2" onClick={ this.previousImage } /> }
+							<IconButton icon="no" aria-label="Remove image" onClick={ this.removeImage } />
```

Comment:
We should also translate the aria-labels

Reply:
As I was typing them, I was thinking, "these need to be translated" and then TOTALLY forgot to do it. Thanks for picking these things up :)